### PR TITLE
Add nodeNumSymmetriesToSample option

### PR DESCRIPTION
**Pull request by roy7**
This also includes all changes from 

https://github.com/lightvector/KataGo/pull/377

### CODE REVIEW

File: cpp/neuralnet/nneval.cpp [hunk 1]
```diff
@@ -473,11 +475,16 @@ void NNEvaluator::serve(
       outputBuf.push_back(emptyOutput);
     }
 
-    int symmetry = defaultSymmetry;
-    if(doRandomize)
-      symmetry = rand.nextUInt(NNInputs::NUM_SYMMETRY_COMBINATIONS);
+    for(int row = 0; row<numRows; row++) {
+      if(buf.resultBufs[row]->symmetry == -1) {
```
Comment (sanderland) [inline cpp/neuralnet/nneval.cpp]: This is hard to follow with the magic -1 number (where is it set, what happens to the value now determined, is there risk of accidentally reusing it?).
When is it not -1 at this point?

File: cpp/configs/analysis_example.cfg [hunk 2]
```diff
@@ -229,8 +229,10 @@ nnRandomize = true
 # The same as chosenMoveSubtract but only prunes moves that fall below the threshold, does not affect moves above
 # chosenMovePrune = 1
 
-# Number of symmetries to sample (WITH replacement) and average at the root
+# Number of symmetries to sample (WITHOUT replacement) and average at the root
 # rootNumSymmetriesToSample = 1
+# Number of symmetries to sample (WITHOUT replacement) and average at interior nodes
```
Comment (sanderland) [inline cpp/configs/analysis_example.cfg]: this slows down the search by this factor, right? should it warn about the effect?

**Review by sanderland — COMMENTED** [review]
few questions

File: cpp/program/setup.cpp [hunk 3]
```diff
@@ -459,6 +459,9 @@ vector<SearchParams> Setup::loadParams(
     if(cfg.contains("rootNumSymmetriesToSample"+idxStr)) params.rootNumSymmetriesToSample = cfg.getInt("rootNumSymmetriesToSample"+idxStr, 1, 16);
     else if(cfg.contains("rootNumSymmetriesToSample"))   params.rootNumSymmetriesToSample = cfg.getInt("rootNumSymmetriesToSample",        1, 16);
     else                                                 params.rootNumSymmetriesToSample = 1;
+    if(cfg.contains("nodeNumSymmetriesToSample"+idxStr)) params.nodeNumSymmetriesToSample = cfg.getInt("nodeNumSymmetriesToSample"+idxStr, 1, 16);
```
Comment (sanderland) [inline cpp/program/setup.cpp]: so 16 here, but 8 below?

File: cpp/neuralnet/nneval.cpp (hunk 1 above)
Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: It's set in the constructor as the default. It is optionally set to specify which symmetry you want. If not set (still -1), then we fall back to using the default or the random as before. Except we're doing this for every row in the array individually now, instead of a single global setting for all rows.

Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: I have a comment in the constructor to document but I can add one here also if you like. Or we could make a global define like SYMMETRY_NOTSPECIFIED to compare against instead of -1.

File: cpp/configs/analysis_example.cfg (hunk 2 above)
Comment (roy7) [reply cpp/configs/analysis_example.cfg]: Yes but it's the same as rootNumSymmetriesToSample which is already in the codebase and doesn't provide a warning. I can add comments about that if you think it's needed. It all defaults to '1' so no effect unless someone intentionally overrides it.

File: cpp/neuralnet/nneval.cpp (hunk 1 above)
Comment (sanderland) [reply cpp/neuralnet/nneval.cpp]: if this if statement checking against -1 wasn't there, what would go wrong?

File: cpp/configs/analysis_example.cfg (hunk 2 above)
Comment (sanderland) [reply cpp/configs/analysis_example.cfg]: Am i understanding correctly that If you ask for 500 visits, with rootsymmetries=4 you'd take 503 (?) nn evals, and with nodesymmetries=4 you'd take 2000 ?

File: cpp/neuralnet/nneval.cpp (hunk 1 above)
Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: It'd probably break. The symmetry has to be between 0 and the max number of symmetries.

Old code was setting it to the default, and then doing randomize, both based on config setting. But this was globally for every eval about to take place in the batch. Now we do the same thing, but on every eval in the batch individually, so we can override some of them if we want to.

The defaults could be added within the constructor for MiscNNInputParams but I was doing it all at once here in the same place the old code used to do it. This also avoids setting it and randomizing it, if it'll just get overwritten later.

File: cpp/configs/analysis_example.cfg (hunk 2 above)
Comment (roy7) [reply cpp/configs/analysis_example.cfg]: Mentally I was thinking rootsymmetries=4 meant the children of the root, but I guess it actually refers to the single root node itself. So yes, rootsymmetries=4 would be 503 nn evals in that case. However the symmetry evals might be counted individually, and so we'd still only do 500 total, with 3 less nodes explored.

File: cpp/neuralnet/nneval.cpp (hunk 1 above)
Comment (sanderland) [reply cpp/neuralnet/nneval.cpp]: No, I mean is it ALWAYS -1 here, such that the check is redundant (it should just set it)

File: cpp/program/setup.cpp (hunk 3 above)
Comment (roy7) [reply cpp/program/setup.cpp]: I just copy/pasted the existing code without thinking about it any. @lightvector, should these two values getInt() with a max value of 8 instead of 16?

Comment (sanderland) [reply cpp/program/setup.cpp]: A higher upper bound makes some sense for with replacements, but yes it seems it should be NNInputs::NUM_SYMMETRY_COMBINATIONS here

File: cpp/neuralnet/nneval.cpp (hunk 1 above)
Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: It gets set at places like https://github.com/roy7/KataGo/pull/1/files#diff-bd67f2c1c385d731233befc2b93c3ad19de564e35748f56424a5eef445e6d829R1139 if you have a specific symmetry you want.

Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: And https://github.com/roy7/KataGo/pull/1/files#diff-d20a00509040ee72ce58635f85b6ca9fcabf064b3a7178f808aecbfe5e3a9b45R2012

Comment (sanderland) [reply cpp/neuralnet/nneval.cpp]: I see, so it kind of belongs to that 'else' branch, but that would lead to some duplication, so it's here via some -1 value that indicates that num symmetries=1.

File: cpp/program/setup.cpp (hunk 3 above)
Comment (roy7) [reply cpp/program/setup.cpp]: Right, with replacement I could see pulling more than 8. If we are getting rid of replacement, then we should max at NNInputs::NUM_SYMMETRY_COMBINATIONS.

File: cpp/neuralnet/nneval.cpp (hunk 1 above)
Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: The # of symmetries is set in the config file. This point in the code has no idea how many we're using, that isn't the purpose here, it's just we're about to do a nn eval and need to know which symmetry you want.

Like if you are in gtp and do 'kata-raw-nn 5' it will give you symmetry 5. buf.resultBufs[row]->symmetry would be '5' in that case so the == -1 test skips it.

Comment (roy7) [reply cpp/neuralnet/nneval.cpp]: Sander, I've replaced -1 with a clear define instead.

**Comment by roy7** [comment]
Various changed in place due to feedback.

**Comment by roy7** [comment]
Note to myself.

If # of symmetries is passed into getHash() then we could do:

```
nnInputParams.numSymmetriesToSample = numSymmetriesToSample;
Hash128 nnHash = NNInputs::getHash(thread.board, thread.history, thread.pla, nnInputParams);
if(numSymmetriesToSample > 1 && (skipCache || nnEvaluator->nnCacheTable == NULL || !nnEvaluator->nnCacheTable->get(nnHash,thread.nnResultBuf))) {
```

and

```
node.nnOutput = std::shared_ptr<NNOutput>(new NNOutput(ptrs));
node.nnOutput.nnHash = nnHash;
if(nnEvaluator->nnCacheTable != NULL) nnEvaluator->nnCacheTable->set(node.nnOutput);
```

In order to cache the averaged result.

File: cpp/configs/analysis_example.cfg (hunk 2 above)
Comment (roy7) [reply cpp/configs/analysis_example.cfg]: I added a performance impact warning

**Comment by roy7** [comment]
In order to access nnCacheTable I moved the symmetry processing into nneval.

I think initial cache logic is all set up, I just need to add the code to use numSymmetriesToSample when calculating the hash.

**Comment by roy7** [comment]
@sanderland Would you like to look it over again now that cache support is added?

File: cpp/neuralnet/nninputs.h [hunk 4]
```diff
@@ -25,23 +25,34 @@ namespace NNPos {
...
+  int symmetry = NNInputs::SYMMETRY_NOTSPECIFIED;
+  // This will get set in initNodeNNOutput if needed, otherwise default to 1 to keep existing behavior within getHash()
+  int numSymmetriesToSample = 1;
 
   static const Hash128 ZOBRIST_CONSERVATIVE_PASS;
   static const Hash128 ZOBRIST_PLAYOUT_DOUBLINGS;
   static const Hash128 ZOBRIST_NN_POLICY_TEMP;
   static const Hash128 ZOBRIST_AVOID_MYTDAGGER_HACK;
+
+  static const Hash128 ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH[NNInputs::NUM_SYMMETRY_COMBINATIONS];
```
Comment (sanderland) [inline cpp/neuralnet/nninputs.h:52]: this has _HASH while the others don't, inconsistent?

File: cpp/search/search.cpp [hunk 5]
```diff
@@ -2001,28 +2001,26 @@ void Search::initNodeNNOutput(
...
-      bool skipCacheThisIteration = skipCache || i > 0; //Skip cache on subsequent iterations to get new random draws for orientation
-      nnEvaluator->evaluate(
-        thread.board, thread.history, thread.pla,
-        nnInputParams,
-        thread.nnResultBuf, skipCacheThisIteration, includeOwnerMap
-      );
-      ptrs.push_back(std::move(thread.nnResultBuf.result));
-    }
-    node.nnOutput = std::shared_ptr<NNOutput>(new NNOutput(ptrs));
+  // We will use numSymmetriesToSample when conputing the hash so that multiple searchers with different settings won't use a cached
```
Comment (sanderland) [inline cpp/search/search.cpp]: computing

File: cpp/search/search.cpp [hunk 6]
```diff
@@ -2001,28 +2001,26 @@ void Search::initNodeNNOutput(
...
+  // eval that had the wrong numSymmetriesToSample
+  nnInputParams.numSymmetriesToSample = isRoot ? searchParams.rootNumSymmetriesToSample : searchParams.nodeNumSymmetriesToSample;
+  if(nnInputParams.numSymmetriesToSample > 1 || skipCache) {
```
Comment (sanderland) [inline cpp/search/search.cpp]: skipCache leading to evaluate_symmetry is weird, I would expect no || here and the skipCache for evaluate to handle it

File: cpp/neuralnet/nninputs.cpp [hunk 7]
```diff
@@ -42,6 +42,17 @@ const Hash128 MiscNNInputParams::ZOBRIST_NN_POLICY_TEMP =
...
+  Hash128(0xe965ae888902aa04ULL, 0x66041fa5e89710aeULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=0
+  Hash128(0xcd7c98236f7734b0ULL, 0xeedb76937b40d0daULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=1
+  Hash128(0x330cc8cd7a4d2b99ULL, 0x66aa846ea74a422bULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=2
+  Hash128(0x60a91a4f39c07263ULL, 0xbeb5d00579ee47f5ULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=3
+  Hash128(0xe22da1a2772dea73ULL, 0xe15afe4705db76f0ULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=4
+  Hash128(0xc06ab8b3ed35edebULL, 0xe60c9cd398808fb8ULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=5
+  Hash128(0x6ddba535ece83fc8ULL, 0x861f5f7cc1d0712cULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=6
+  Hash128(0x4c9153b26848d715ULL, 0xc1dc7c20fce60760ULL),  //Based on sha256 hash of MiscNNInputParams::ZOBRIST_NUM_SYMMETRIES_TO_SAMPLE_HASH=7
```
Comment (sanderland) [inline cpp/neuralnet/nninputs.cpp:53]: haha is this really how these hash values are determined. such math :D

File: cpp/neuralnet/nneval.cpp [hunk 8]
```diff
@@ -948,6 +959,48 @@ void NNEvaluator::evaluate(
...
+  Player nextPlayer,
+  MiscNNInputParams& nnInputParams,
+  NNResultBuf& buf,
+  bool skipCache,
+  bool includeOwnerMap,
+  Rand& rand
+) {
+  assert(nnInputParams.numSymmetriesToSample > 1);
+  Hash128 nnHash = NNInputs::getHash(board, history, nextPlayer, nnInputParams);
+  if(skipCache || nnCacheTable == NULL || !nnCacheTable->get(nnHash,buf.result)) {
```
Comment (sanderland) [inline cpp/neuralnet/nneval.cpp:977]: I don't like the nnCache lookup in this if statement, it's too complex. and not clear why after getting the cache it still needs to call evaluate. Can it just be a block above it if(not skipcache and cachehit) return

File: cpp/neuralnet/nninputs.h (hunk 4 above)
Comment (roy7) [reply cpp/neuralnet/nninputs.h:52]: I found that all other uses of ZOBRIST in the current code use the name _HASH when it is an array instead of a single value. So I followed that example.

File: cpp/search/search.cpp (hunk 5 above)
Comment (roy7) [reply cpp/search/search.cpp]: Fixed.

File: cpp/search/search.cpp (hunk 6 above)
Comment (roy7) [reply cpp/search/search.cpp]: That's a good point. Because we don't check skipCache below in the normal evaluate() call. It might be an artifact of the told test against "i > 0". So just check here for >1 and do the if/else and pass skipCache in the function like usual.

In fact, num == 1 and skipCache true would call wrong function. Thanks.

File: cpp/neuralnet/nninputs.cpp (hunk 7 above)
Comment (roy7) [reply cpp/neuralnet/nninputs.cpp:53]: Amazing isn't it. ;)

File: cpp/neuralnet/nneval.cpp (hunk 8 above)
Comment (roy7) [reply cpp/neuralnet/nneval.cpp:977]: Because we're only calling evaluate on a cache miss. If it's in the cache, !get() will be false.

Comment (roy7) [reply cpp/neuralnet/nneval.cpp:977]: I could add a comment above it if that helps?

Comment (sanderland) [reply cpp/neuralnet/nneval.cpp:977]: yes, the !get will be false, and it will go to else to evaluate, even though it's already in buf, right?

Comment (roy7) [reply cpp/neuralnet/nneval.cpp:977]: Err right. What's going on there. I need to review it when I have a few minutes.

The logic maybe should be a 2nd check after this one that if it's in the buf already don't eval at all, just return? Or is there no harm in calling evaluate() if it's already in buf, since if it's cached there's no nn eval wasted? But yeah extra processing steps we can skip we should skip.

Is it possible for cache to hit but ownership map need updated? If I recall evaluate() does more than just fetch the cache.

**Comment by roy7** [comment]
Moving PR upstream.
